File: leetcode/hard/295_find_median_from_data_stream.py
# -----------------------------------------
# My Solution
#
# Time  Complexity: O(nlog(n))
# Space Complexity: O(n)
# -----------------------------------------
import bisect

class MedianFinder:

    def __init__(self):
        self.arr = []


    def addNum(self, num: int) -> None:
        if len(self.arr) == 0:
            self.arr.append(num)
        else:
            bisect.insort(self.arr, num)


    def findMedian(self) -> float:
        if len(self.arr) % 2 == 1:
            return self.arr[len(self.arr) // 2]
        else:
            return (self.arr[len(self.arr) // 2] + self.arr[len(self.arr) // 2 - 1]) / 2

# -----------------------------------------
# Model Solution: Two Heaps
#
# Time  Complexity: O(log(n))
# Space Complexity: O(n)
# -----------------------------------------
# Ref: https://www.youtube.com/watch?v=U_tisGQrQ3c
import heapq

class MedianFinder:

    def __init__(self):
        self.max_heap = []
        self.min_heap = []


    def addNum(self, num: int) -> None:
        if len(self.max_heap) == 0 or -self.max_heap[0] > num:
            heapq.heappush(self.max_heap, -num)
        else:
            heapq.heappush(self.min_heap, num)

        if len(self.max_heap) > len(self.min_heap) + 1:
            val = -heapq.heappop(self.max_heap)
            heapq.heappush(self.min_heap, val)
        elif len(self.max_heap) < len(self.min_heap):
            val = heapq.heappop(self.min_heap)
            heapq.heappush(self.max_heap, -val)


    def findMedian(self) -> float:
        if (len(self.min_heap) + len(self.max_heap)) % 2 == 0:
            return (-self.max_heap[0] + self.min_heap[0]) / 2
        else:
            return -self.max_heap[0]

File: leetcode/hard/295_find_median_from_data_stream.cpp
/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

// -----------------------------------------
// Model Solution: Multiset
//
// Time  Complexity: O(log(n))
// Space Complexity: O(n)
// -----------------------------------------
// n := nums.size()
//
// Ref: https://leetcode.com/problems/find-median-from-data-stream/discuss/1330697/C%2B%2B-Short-and-Clean-Solution-Using-One-Multiset-5-Lines-of-Code
class MedianFinder {
public:
    /** initialize your data structure here. */
    MedianFinder() {}

    void addNum(int num) {
        nums.insert(num);
        nums_size_even = !nums_size_even;

        // mid_iterator stays at nums.end() after first element inserted and ++mid_iterator moves mid_iterator to nums.begin()
        if (mid_iterator == nums.end() || (nums_size_even && *mid_iterator <= num))
            ++mid_iterator;
        if (!nums_size_even && *mid_iterator > num)
            --mid_iterator;
    }

    double findMedian() {
        return nums_size_even ? ((*mid_iterator + *prev(mid_iterator)) / 2.0) : *mid_iterator;
    }

private:
    multiset<int> nums;
    multiset<int>::iterator mid_iterator = nums.begin();
    bool nums_size_even = true;
};
